miden-crypto/src/merkle/sparse_path.rs

```diff
@@ -0,0 +1,385 @@
+    path: SparseMerklePath,
+
+    /// The depth a `next()` call will get. It will only be None if someone calls `next_back()` at
+    /// depth 0, to indicate that all further `next_back()` calls must also return `None`.
+    next_depth: Option<u8>,
+
+    /// "Cached" value of `path.depth()`.
+    tree_depth: u8,
+}
+
+impl Iterator for SparseMerkleIter {
+    type Item = RpoDigest;
+
+    fn next(&mut self) -> Option<RpoDigest> {
+        // If `next_depth` is None, then someone called `next_back()` at depth 0.
+        let next_depth = self.next_depth.unwrap_or(0);
+        if next_depth >= self.tree_depth {
+            return None;
+        }
+
+        match self.path.get(next_depth) {
+            Some(node) => {
+                self.next_depth = Some(next_depth + 1);
+                Some(node)
+            },
+            None => None,
+        }
+    }
+
+    // SparseMerkleIter always knows its exact size.
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let next_depth = self.next_depth.unwrap_or(0);
+        let len: usize = self.path.depth().into();
+        let remaining = len - next_depth as usize;
+        (remaining, Some(remaining))
+    }
+}
+
+impl ExactSizeIterator for SparseMerkleIter {
+    fn len(&self) -> usize {
+        let next_depth = self.next_depth.unwrap_or(0);
+        (self.path.depth() - next_depth) as usize
+    }
+}
+
+impl DoubleEndedIterator for SparseMerkleIter {
+    fn next_back(&mut self) -> Option<RpoDigest> {
+        // While `next_depth` is None, all calls to `next_back()` also return `None`.
+        let next_depth = self.next_depth?;
+
+        match self.path.get(next_depth) {
+            Some(node) => {
+                self.next_depth = if next_depth == 0 { None } else { Some(next_depth - 1) };
+                Some(node)
+            },
+            None => None,
+        }
+    }
+}
+
+// SPARSE MERKLE PATH CONTAINERS
+// ================================================================================================
+/// A container for a [crate::Word] value and its [SparseMerklePath] opening.
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+pub struct SparseValuePath {
+    /// The node value opening for `path`.
+    pub value: RpoDigest,
+    /// The path from `value` to `root` (exclusive), using an efficient memory representation for
+    /// empty nodes.
+    pub path: SparseMerklePath,
+}
+
+impl SparseValuePath {
+    /// Convenience function to construct a [SparseValuePath].
+    ///
+    /// `value` is the value `path` leads to, in the tree.
+    pub fn new(value: RpoDigest, path: SparseMerklePath) -> Self {
+        Self { value, path }
+    }
+}
+
+impl From<(SparseMerklePath, Word)> for SparseValuePath {
+    fn from((path, value): (SparseMerklePath, Word)) -> Self {
+        SparseValuePath::new(value.into(), path)
+    }
+}
+
+// SERIALIZATION
+// ================================================================================================
```

Comment:
I would probably group the serialization for `SparseMerklePath` together with that struct, so maybe before the `SparseMerkleIter`